RecoPixelVertexing/PixelTriplets/plugins/CAHitNtupletHeterogeneousEDProducer.cc

```diff
@@ -0,0 +1,158 @@
+public:
+  CAHitNtupletHeterogeneousEDProducer(const edm::ParameterSet &iConfig);
+  ~CAHitNtupletHeterogeneousEDProducer();
+
+  static void fillDescriptions(edm::ConfigurationDescriptions &descriptions);
+  void beginStreamGPUCuda(edm::StreamID streamId,
+                          cuda::stream_t<> &cudaStream) override;
+
+  void acquireGPUCuda(const edm::HeterogeneousEvent &iEvent,
+                      const edm::EventSetup &iSetup,
+                      cuda::stream_t<> &cudaStream) override;
+  void produceGPUCuda(edm::HeterogeneousEvent &iEvent,
+                      const edm::EventSetup &iSetup,
+                      cuda::stream_t<> &cudaStream) override;
+  void produceCPU(edm::HeterogeneousEvent &iEvent,
+                  const edm::EventSetup &iSetup) override;
+
+private:
+  edm::EDGetTokenT<IntermediateHitDoublets> doubletToken_;
+
+  edm::RunningAverage localRA_;
+  CAHitQuadrupletGeneratorGPU generator_;
+  bool emptyRegionDoublets = false;
+  std::unique_ptr<RegionsSeedingHitSets> seedingHitSets;
+  std::vector<OrderedHitSeeds> ntuplets;
+};
+
+CAHitNtupletHeterogeneousEDProducer::CAHitNtupletHeterogeneousEDProducer(
+    const edm::ParameterSet &iConfig)
+    : HeterogeneousEDProducer<heterogeneous::HeterogeneousDevices<
+          heterogeneous::GPUCuda, heterogeneous::CPU>>(iConfig), doubletToken_(consumes<IntermediateHitDoublets>(iConfig.getParameter<edm::InputTag>("doublets"))),
+      generator_(iConfig, consumesCollector()) {
+  produces<RegionsSeedingHitSets>();
+}
+
+CAHitNtupletHeterogeneousEDProducer::~CAHitNtupletHeterogeneousEDProducer() {
+
+}
+
+void CAHitNtupletHeterogeneousEDProducer::fillDescriptions(
+    edm::ConfigurationDescriptions &descriptions) {
+  edm::ParameterSetDescription desc;
+
+  desc.add<edm::InputTag>("doublets", edm::InputTag("hitPairEDProducer"));
+  CAHitQuadrupletGeneratorGPU::fillDescriptions(desc);
+
+  auto label = CAHitQuadrupletGeneratorGPU::fillDescriptionsLabel() +
+               std::string("EDProducer");
+  descriptions.add(label, desc);
+}
+
+void CAHitNtupletHeterogeneousEDProducer::beginStreamGPUCuda(
+    edm::StreamID streamId, cuda::stream_t<> &cudaStream) {
+
+}
+
+void CAHitNtupletHeterogeneousEDProducer::produceGPUCuda(
+    edm::HeterogeneousEvent &iEvent, const edm::EventSetup &iSetup,
+    cuda::stream_t<> &cudaStream) {
+
+  if (!emptyRegionDoublets) {
+    edm::Handle<IntermediateHitDoublets> hdoublets;
+    iEvent.getByToken(doubletToken_, hdoublets);
+    const auto &regionDoublets = *hdoublets;
+    const SeedingLayerSetsHits &seedingLayerHits =
+        regionDoublets.seedingLayerHits();
+    int index = 0;
+    for (const auto &regionLayerPairs : regionDoublets) {
+      const TrackingRegion &region = regionLayerPairs.region();
+      auto seedingHitSetsFiller = seedingHitSets->beginRegion(&region);
+      generator_.fillResults(regionDoublets, ntuplets, iSetup, seedingLayerHits,
+                             cudaStream.id());
```

Comment:
To me it looks like `CAHitQuadrupletGeneratorGPU::fillResults()` already loops over the regions and fills `ntuplets` for all regions. Why does it have to be repeated here for each region? Should the loop perhaps be something along
```cpp
  generator_.fillResults(regionDoublets, ntuplets, iSetup, seedingLayerHits, cudaStream.id());
  int index = 0;
  for (const auto &regionLayerPairs : regionDoublets) {
    const TrackingRegion &region = regionLayerPairs.region();
    auto seedingHitSetsFiller = seedingHitSets->beginRegion(&region);
    fillNtuplets(seedingHitSetsFiller, ntuplets[index]);
    index++;
  )
```
instead?